src/security/ddwaf_memres.h

```diff
@@ -0,0 +1,99 @@
+#pragma once
+
+#include <ddwaf.h>
+
+#include <algorithm>
+#include <cstdint>
+#include <cstdlib>
+#include <memory>
+#include <stdexcept>
+#include <vector>
+
+extern "C" {
+#include <ngx_core.h>
+}
+
+namespace datadog::nginx::security {
+
+class DdwafMemres {
+  static inline constexpr std::size_t kMinObjSegSize = 20;
+  static inline constexpr std::size_t kMinStrSegSize = 512;
+
+ public:
+  DdwafMemres() = default;
+  DdwafMemres(const DdwafMemres &) = delete;
+  DdwafMemres &operator=(const DdwafMemres &) = delete;
+  DdwafMemres(DdwafMemres &&) = default;
+  DdwafMemres &operator=(DdwafMemres &&) = default;
+  ~DdwafMemres() = default;
+
+  template <typename T = ddwaf_object>
+  T *allocate_objects(std::size_t num_objects) {
```

Comment:
Why is this method templated? Is it to alias to one of your object variants? Would be nice to have a comment somewhere, rationale isn't immediately obvious.

Reply:
Yes, that's the idea, as reflected in the static assertions (T must be a subclass of ddwaf_object, with same size and alignment and must satisfy the standard layout requirements), although I see calling code only uses ddwaf_object and `dnsec::ddwaf_obj`, so I could also return the latter.